### PR TITLE
[camera][web] Rewrite in hooks

### PR DESCRIPTION
**Pull request by EvanBacon**
# Why

- There were a few mounting, and state changing bugs that could be more elegantly fixed using hooks.
  - Swapping cameras would flash even if the direction didn't change
  - Swapping cameras wouldn't set the defined capabilities (i.e. torch mode, auto focus on Android).
- Using hooks will make it easier to extend the interface and add support for barcode scanning.
- Code was just generally really messy

# Test Plan

- Demo: https://expo-web-9558.now.sh/components/camera

- Fixed the NCL camera screen on web
- Fixed a bug where webpack 5 doesn't observe changes to `packages/`
- Tested on Android, iOS, and desktop (macos).

### CODE REVIEW

File: apps/native-component-list/webpack.config.js [hunk 1]
```diff
@@ -0,0 +1,10 @@
+const createConfigAsync = require('@expo/webpack-config');
+
+module.exports = async env => {
+  const config = await createConfigAsync(env);
+  // allow reloading when the packages/ are updated.
```
Comment (lukmccall) [inline apps/native-component-list/webpack.config.js:5]:
```suggestion
  // allow reloading when the packages are updated.
```

File: packages/expo-camera/src/CameraModule/CameraUtils.ts [hunk 2]
```diff
@@ -171,3 +178,199 @@ export async function getStreamDevice(
 export function isWebKit(): boolean {
   return /WebKit/.test(navigator.userAgent) && !/Edg/.test(navigator.userAgent);
 }
+
+export function compareStreams(a: MediaStream | null, b: MediaStream | null): boolean {
+  if (!a || !b) return false;
```
Comment (lukmccall) [inline packages/expo-camera/src/CameraModule/CameraUtils.ts:183]:
```suggestion
  if (!a || !b) {
    return false;
  }
```

File: packages/expo-camera/src/CameraModule/CameraUtils.ts [hunk 3]
```diff
@@ -171,3 +178,199 @@ export async function getStreamDevice(
...
+
+  if (capabilities.focusMode && settings.autoFocus !== undefined) {
+    constraints.focusMode = _validatedConstrainedValue(
+      'focusMode',
+      'autoFocus',
+      CapabilityUtils.convertAutoFocusJSONToNative
+    );
+  }
+
+  console.log('can use torch: ', capabilities.torch);
```
Comment (lukmccall) [inline packages/expo-camera/src/CameraModule/CameraUtils.ts:278]:
```suggestion
```

File: packages/expo-camera/src/CameraModule/CameraUtils.ts [hunk 4]
```diff
@@ -171,3 +178,199 @@ export async function getStreamDevice(
...
+
+  if (capabilities.whiteBalanceMode && settings.whiteBalance !== undefined) {
+    constraints.whiteBalanceMode = _validatedConstrainedValue(
+      'whiteBalanceMode',
+      'whiteBalance',
+      CapabilityUtils.convertWhiteBalanceJSONToNative
+    );
+  }
+
+  console.log('apply constraints: ', constraints);
```
Comment (lukmccall) [inline packages/expo-camera/src/CameraModule/CameraUtils.ts:296]:
```suggestion
```

File: packages/expo-camera/src/CameraModule/CameraUtils.ts [hunk 5]
```diff
@@ -171,3 +178,199 @@ export async function getStreamDevice(
...
+  await track.applyConstraints({ advanced: [constraints] as any });
+}
+
+export function stopMediaStream(stream: MediaStream | null) {
+  if (!stream) return;
+  if (stream.getAudioTracks) stream.getAudioTracks().forEach(track => track.stop());
+  if (stream.getVideoTracks) stream.getVideoTracks().forEach(track => track.stop());
+  if (isMediaStreamTrack(stream)) stream.stop();
```
Comment (lukmccall) [inline packages/expo-camera/src/CameraModule/CameraUtils.ts:304]:
```suggestion
  if (!stream) {
    return;
  }
  if (stream.getAudioTracks) {
    stream.getAudioTracks().forEach(track => track.stop());
  }
  if (stream.getVideoTracks) { 
    stream.getVideoTracks().forEach(track => track.stop());
  }
  if (isMediaStreamTrack(stream)) { 
    stream.stop();
  }
```

File: packages/expo-camera/src/CameraModule/CameraUtils.ts [hunk 6]
```diff
@@ -171,3 +178,199 @@ export async function getStreamDevice(
...
+
+  return false;
+}
+
+function isMediaStreamTrack(input: any): input is MediaStreamTrack {
+  return typeof input.stop === 'function';
+}
+
+function convertNormalizedSetting(range: MediaSettingsRange, value?: number): number | undefined {
+  if (!value) return;
```
Comment (lukmccall) [inline packages/expo-camera/src/CameraModule/CameraUtils.ts:345]:
```suggestion
  if (!value) { 
    return;
  }
```

File: packages/expo-camera/src/CameraModule/CameraUtils.ts [hunk 7]
```diff
@@ -171,3 +178,199 @@ export async function getStreamDevice(
...
+}
+
+function convertRange(value: number, r2: number[], r1: number[] = [0, 1]): number {
+  return ((value - r1[0]) * (r2[1] - r2[0])) / (r1[1] - r1[0]) + r2[0];
+}
+
+function validatedConstrainedValue(
+  constraintKey: string,
+  settingsKey: string,
+  convertedSetting: any,
```
Comment (lukmccall) [inline packages/expo-camera/src/CameraModule/CameraUtils.ts:359]: Can we narrow this type?

**Review by lukmccall — COMMENTED** [review]
It's looking good 💪
I've left some comments ;)

File: packages/expo-camera/src/CameraModule/CameraUtils.ts [hunk 8]
```diff
@@ -171,3 +178,199 @@ export async function getStreamDevice(
...
+    'zoom',
+  ];
+
+  for (const property of clampedValues) {
+    if (capabilities[property]) {
+      constraints[property] = convertNormalizedSetting(capabilities[property], settings[property]);
+    }
+  }
+
+  const _validatedConstrainedValue = (key, propName, converter) =>
```
Comment (bbarthec) [inline packages/expo-camera/src/CameraModule/CameraUtils.ts:260]:
```suggestion
  const _validatedConstrainedValue = (key: string, propName: string, converter: (value: any) => any) =>
```
I'm not a fun of these `any` in `converter` 😞

File: packages/expo-camera/src/CameraModule/CameraUtils.ts [hunk 9]
```diff
@@ -171,3 +178,199 @@ export async function getStreamDevice(
...
+  if (stream instanceof MediaStream) {
+    const videoTrack = stream.getVideoTracks()[0];
+
+    if (typeof videoTrack.getCapabilities === 'undefined') {
+      return false;
+    }
+
+    const capabilities: any = videoTrack.getCapabilities();
+
+    return !!capabilities[keyName];
```
Comment (bbarthec) [inline packages/expo-camera/src/CameraModule/CameraUtils.ts:334]:
```suggestion
    return Boolean(videoTrack.getCapabilities.?()[keyName]);
```

File: packages/expo-camera/src/ExponentCamera.web.tsx [hunk 10]
```diff
@@ -1,144 +1,90 @@
...
+export type ExponentCameraRef = {
+  getAvailablePictureSizes: (ratio: string) => Promise<string[]>;
+  takePicture: (options: CameraPictureOptions) => Promise<CameraCapturedPicture>;
+  resumePreview: () => Promise<void>;
+  pausePreview: () => Promise<void>;
+};
+
+const ExponentCamera = React.forwardRef(
+  (
+    { type, pictureSize, ...props }: CameraNativeProps & { children?: any },
```
Comment (bbarthec) [inline packages/expo-camera/src/ExponentCamera.web.tsx:19]:
```suggestion
    { type, pictureSize, ...props }: CameraNativeProps & { children?: React.ReactNode },
```

**Review by bbarthec — COMMENTED** [review]
LGTM 👍 
Nice work here. It definitely looks cleaner now 🎉

**Review by lukmccall — APPROVED** [review]

File: packages/expo-camera/src/useWebCameraStream.ts [hunk 11]
```diff
@@ -0,0 +1,191 @@
...
+
+  React.useEffect(() => {
+    resumeAsync();
+  }, [preferredType]);
+
+  React.useEffect(() => {
+    const changes: WebCameraSettings = {};
+
+    for (const key of Object.keys(settings)) {
+      if (!VALID_SETTINGS_KEYS.includes(key)) continue;
```
Comment (bbarthec) [inline packages/expo-camera/src/useWebCameraStream.ts:98]:
```suggestion
      if (!VALID_SETTINGS_KEYS.includes(key)) {
        continue;
      }
```

File: packages/expo-camera/src/WebCameraUtils.ts [hunk 12]
```diff
@@ -0,0 +1,393 @@
...
+
+  const base64 = toDataURL(canvas, imageType!, quality);
+  return base64;
+}
+
+function getSupportedConstraints(): MediaTrackSupportedConstraints | null {
+  if (navigator.mediaDevices && navigator.mediaDevices.getSupportedConstraints) {
+    return navigator.mediaDevices.getSupportedConstraints();
+  }
+  return null;
```
Comment (bbarthec) [inline packages/expo-camera/src/WebCameraUtils.ts:118]:
```suggestion
  return navigator.mediaDevices?.getSupportedConstraints?.() ?? null;
```

**Review by bbarthec — APPROVED** [review]
In general it looks good to me 👍 
The only important thing is that enum naming convention.

File: packages/expo-camera/src/WebCameraUtils.ts (hunk 12 above)
Comment (EvanBacon) [reply packages/expo-camera/src/WebCameraUtils.ts:118]: https://github.com/expo/expo/pull/9558/commits/6cfdf42bf3bcb9ea0a7e826604d9fe3c4698868a

File: packages/expo-camera/src/ExponentCameraManager.web.ts [hunk 13]
```diff
@@ -42,26 +45,35 @@ export default {
...
+    return types.filter(Boolean) as string[];
   },
-  async getAvailablePictureSizes(ratio: string, camera: ExponentCamera): Promise<string[]> {
+  async getAvailablePictureSizes(ratio: string, camera: ExponentCameraRef): Promise<string[]> {
     return await camera.getAvailablePictureSizes(ratio);
   },
+
+  // TODO(Bacon): Is video possible?
+  // record(options): Promise
+  // stopRecording(): Promise<void>
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo-camera/src/ExponentCameraManager.web.ts:78]: I think for now a comment is the correct place for this.

File: packages/expo-camera/src/Camera.types.ts [hunk 14]
```diff
@@ -5,12 +5,61 @@ import {
   PermissionExpiration,
 } from 'unimodules-permissions-interface';
 
+export enum CameraType {
+  front = 'front',
+  back = 'back',
+}
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo-camera/src/Camera.types.ts:11]: maybe we could make these breaking changes in another PR

File: packages/expo-camera/src/Camera.types.ts [hunk 15]
```diff
@@ -5,12 +5,61 @@ import {
...
+
+export enum ImageType {
+  png = 'png',
+  jpg = 'jpg',
+}
+
+export type ImageParameters = {
+  imageType: ImageType;
+  quality: number | null;
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo-camera/src/Camera.types.ts:20]: https://github.com/expo/expo/pull/9558#discussion_r466626062

File: packages/expo-camera/src/Camera.types.ts [hunk 16]
```diff
@@ -5,12 +5,61 @@ import {
...
 } from 'unimodules-permissions-interface';
 
+export enum CameraType {
+  front = 'front',
+  back = 'back',
+}
+
+export enum ImageType {
+  png = 'png',
+  jpg = 'jpg',
```
[the remark this answers is not in the document]
Comment (EvanBacon) [reply packages/expo-camera/src/Camera.types.ts:15]: https://github.com/expo/expo/pull/9558#discussion_r466626062

**Pull request merged**